microsoft/testsuites/dpdk/dpdksuite.py

```diff
@@ -257,4 +380,85 @@ def initialize_node_resources(node: Node, log: Logger) -> NodeResources:
     assert_that(len(node_nic_info)).described_as(
         "Test needs at least 2 NICs on the test node."
     ).is_greater_than_or_equal_to(2)
-    return NodeResources(node_nic_info, testpmd)
+    return NodeResources(node, node_nic_info, testpmd)
+
+
+def _threaded_resource_init(
+    node: Node, log: Logger, output: List[NodeResources]
+) -> None:
+    # threading.Thread target function, results aggregated into output list
+    output.append(initialize_node_resources(node, log))
+
+
+def _threaded_testpmd(
+    node_cmd_pair: Tuple[NodeResources, str],
+    seconds: int,
+    log: Logger,
+    output: Dict[NodeResources, str],
+) -> None:
+    testkit, cmd = node_cmd_pair
+    output[testkit] = testkit.testpmd.run_for_n_seconds(cmd, seconds)
+
+
+def _run_testpmd_concurrent(
+    node_cmd_pairs: Dict[NodeResources, str],
+    seconds: int,
+    log: Logger,
+) -> Dict[NodeResources, str]:
+    output: Dict[NodeResources, str] = dict()
+    threads: List[threading.Thread] = []
+    for pair in node_cmd_pairs.items():
+        threads.append(
+            threading.Thread(
+                target=_threaded_testpmd, args=(pair, seconds, log, output)
+            )
+        )
+    for thread in threads:
+        thread.start()
+    threads.reverse()
+    for thread in threads:
+        thread.join()
+    return output
```

Comment:
The `parallel.py` has a concurrent implementation, it can be reused here. You can expose the `return_when` in the `wait_worker` to support `FIRST_COMPLETED` and `ALL_COMPLETED`.

Reply:
I don't see a file named parallel.py in lisa, is this an external module or something?